clients/roscpp/src/libros/service_server_link.cpp

```diff
@@ -362,15 +368,38 @@ bool ServiceServerLink::call(const SerializedMessage& req, SerializedMessage& re
     processNextCall();
   }
 
+  bool timed_out = false;
   {
     boost::mutex::scoped_lock lock(info->finished_mutex_);
 
+    using namespace boost::chrono;
+    steady_clock::time_point end_time = steady_clock::now() + nanoseconds(timeout.toNSec());
+
     while (!info->finished_)
     {
-      info->finished_condition_.wait(lock);
+      if (timeout > ros::WallDuration(0.0))
+      {
+        if (info->finished_condition_.wait_until(lock, end_time) == boost::cv_status::timeout)
+        {
+          ROS_WARN("Service [%s] call failed: no response for %fsec", service_name_.c_str(), timeout.toSec());
+          timed_out = true;
+          break;
+        }
+      }
+      else
+      {
+        info->finished_condition_.wait(lock);
+      }
     }
   }
 
+  if (timed_out)
+  {
+    // TODO: actually properly cancel the request (connection read)
```

Comment:
What is the status of this TODO?

Reply:
I'm actually unsure about the internal workings here, but as far as I can see `callFinished` and the likes only remove the callback but don't really "cancel" the actual request.
Not even sure if that is possible.
Notifying the service call provider is probably out of the scope anyway.